signer/src/json.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright 2019-2024 Parity Technologies (UK) Ltd.
+// This file is dual-licensed as Apache-2.0 or GPL-3.0.
+// see LICENSE for license details.
+
+//! A Polkadot-JS account loader.
+
+use base64::Engine;
+use core::fmt::Display;
+use crypto_secretbox::{
+    aead::{Aead, KeyInit},
+    Key, Nonce, XSalsa20Poly1305,
+};
+use serde::Deserialize;
+use subxt_core::utils::AccountId32;
+
+use crate::sr25519;
+
+/// Error
+#[derive(Debug)]
+pub enum Error {
+    /// The keypair has an unsupported encoding.
+    UnsupportedEncoding,
+    /// Base64 decoding error.
+    Base64(base64::DecodeError),
+    /// Wrong Scrypt parameters
+    UnsupportedScryptParameters {
+        /// N
+        n: u32,
+        /// p
+        p: u32,
+        /// r
+        r: u32,
+    },
+    /// Decryption error.
+    Secretbox(crypto_secretbox::Error),
+    /// The decrypted keys are not valid.
+    InvalidKeys,
+}
+
+impl_from!(base64::DecodeError => Error::Base64);
+impl_from!(crypto_secretbox::Error => Error::Secretbox);
+
+impl Display for Error {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            Error::UnsupportedEncoding => write!(f, "Unsupported encoding."),
+            Error::Base64(e) => write!(f, "Base64 decoding error: {e}"),
+            Error::UnsupportedScryptParameters { n, p, r } => {
+                write!(f, "Unsupported Scrypt parameters: N: {n}, p: {p}, r: {r}")
+            }
+            Error::Secretbox(e) => write!(f, "Decryption error: {e}"),
+            Error::InvalidKeys => write!(f, "The decrypted keys are not valid."),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for Error {}
+
+#[derive(Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+struct EncryptedJsonDescriptor {
+    /// Descriptor for the content
+    content: Vec<String>,
+    /// The encoding (in current/latest versions this is always an array)
+    r#type: Vec<String>,
+    /// The version of encoding applied
+    version: String,
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+#[allow(dead_code)]
+struct Meta {
+    genesis_hash: String,
+    is_hardware: Option<bool>,
+    name: String,
+    tags: Option<Vec<String>>,
+    when_created: u64,
+}
+
+/// https://github.com/polkadot-js/common/blob/37fa211fdb141d4f6eb32e8f377a4651ed2d9068/packages/keyring/src/types.ts#L67
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+#[allow(dead_code)]
+pub struct KeyringPairJson {
+    /// The encoded string
+    encoded: String,
+    /// The encoding used
+    encoding: EncryptedJsonDescriptor,
+    /// The ss58 encoded address or the hex-encoded version (the latter is for ETH-compat chains)
+    address: AccountId32,
+    /// The underlying metadata associated with the keypair
+    meta: Meta,
+}
+
+impl KeyringPairJson {
+    /// Decrypt JSON keypair.
+    pub fn decrypt(self, password: &str) -> Result<sr25519::Keypair, Error> {
```

Comment:
I reckon that the API in this module would be cleaner if the public API for it looked something like this:

```rust
mod polkadot_js_compat {
    /// Given a JSON keypair as exported from Polkadot-JS, this returns an [`sr25519::Keypair`]
    pub fn decrypt_json(json: &str, password: &str) -> Result<sr25519::Keypair, Error> { .. }

    /// Possible errors when calling the [`from_json`] function.
    pub enum Error { .. }
}
```

That way there is no visible struct and just one method that takes the JSON string and gives back a keypair or an error.

Reply:
Implemented.